mesa/time.py

```diff
@@ -327,3 +329,111 @@ def get_type_count(self, type_class: type[Agent]) -> int:
         """
         return len(self.agents_by_type[type_class])
+
+
+class DiscreteEventScheduler(BaseScheduler):
+    """
+    A scheduler for discrete event simulation in Mesa.
+
+    This scheduler manages events where each event is associated with a
+    specific time and agent. The scheduler advances time not in fixed
+    increments, but to the moment the next event is scheduled to occur.
+
+    This implementation uses a priority queue (heapq) to manage events. Each
+    event is a tuple of the form (time, random_value, agent), where:
+        - time (float): The scheduled time for the event.
+        - random_value (float): A secondary sorting criterion to randomize
+          the order of events that are scheduled for the same time.
+        - agent (Agent): The agent associated with the event.
+
+    The random value for secondary sorting ensures that when two events are
+    scheduled for the same time, their execution order is randomized, thus
+    preventing direct comparison issues between different types of agents and
+    maintaining the integrity of the simulation's randomness.
+
+    Attributes:
+        model (Model): The model instance associated with the scheduler.
+        event_queue (list): A priority queue of scheduled events.
+        time_step (int or float): The fixed time period by which the model advances
+                                  on each step. Defaults to 1.
+
+    Methods:
+        schedule_event(time, agent): Schedule an event for a specific time.
+        schedule_in(delay, agent): Schedule an event after a specified delay.
+        step(): Execute all events within the next time_step period.
+        get_next_event_time(): Returns the time of the next scheduled event.
+
+    Usage:
+        1. Instantiate the DiscreteEventScheduler with a model instance and a time_step period.
+        2. Add agents to the scheduler using schedule.add(). With schedule_now=True (default),
+              the first event for the agent will be scheduled immediately.
+        3. In the Agent step() method, schedule the next event for the agent
+              (using schedule_in or schedule_event).
+        3. Add self.schedule.step() to the model's step() method, as usual.
+
+    Now, with each model step, the scheduler will execute all events within the
+    next time_step period, and advance time one time_step forward.
+    """
+
+    def __init__(self, model: Model, time_step: TimeT = 1) -> None:
+        super().__init__(model)
+        self.event_queue: list[tuple[TimeT, float, Agent]] = []
+        self.time_step: TimeT = time_step  # Fixed time period for each step
+
+    def schedule_event(self, time: TimeT, agent: Agent) -> None:
+        """Schedule an event for an agent at a specific time."""
+        if time < self.time:
+            raise ValueError(f"Scheduled time ({time}) must be >= the current time ({self.time})")
+        event = (
+            time,
+            random.random(),
```

Comment:
I think this needs to be `self.model.random.random()`. Because the random source could be from NumPy, and has been seeded. Hardcoding it to `random.random` is a gotcha that may cause non-deterministic behavior.

Reply:
Updated